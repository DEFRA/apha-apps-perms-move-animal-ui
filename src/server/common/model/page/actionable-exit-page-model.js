import { ExitPage } from './exit-page-model.js'

/** @import { AnswerModel } from '../answer/answer-model.js' */
/** @import { Page } from './page-model.js' */

/** @typedef {{ page: Page, answer: AnswerModel}}  IndirectActionConfig */

export class ActionableExitPage extends ExitPage {
  /** @type {IndirectActionConfig} */
  _indirectAction

  /** @type {IndirectActionConfig} */
  get indirectAction() {
    return this._indirectAction
  }
}
