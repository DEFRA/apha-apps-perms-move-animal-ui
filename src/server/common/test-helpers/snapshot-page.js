import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
