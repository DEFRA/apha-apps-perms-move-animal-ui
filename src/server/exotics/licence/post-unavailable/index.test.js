import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { ActionableExitPage } from '~/src/server/common/model/page/actionable-exit-page-model.js'
import {
  emailOrPostPage,
  Answer as EmailOrPostAnswer
} from '../email-or-post/index.js'
import { postUnavailablePage } from './index.js'
import { EmailPage } from '../email/index.js'

const sectionKey = 'licence'
const pageUrl =
  '/exotics/receiving-the-licence/select-post-can-not-use-this-service'
const page = postUnavailablePage
const question = 'This service does not currently send licences by post'
const view = `exotics/licence/post-unavailable/index`

describe('PostUnavailablePage', () => {
  it('should be an instance of ActionableExitPage', () => {
    expect(page).toBeInstanceOf(ActionableExitPage)
  })

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct page title', () => {
    expect(page.pageTitle).toBe(question)
  })

  describe('indirectAction', () => {
    it('should return an object with page set to cphNumberKnownPage', () => {
      expect(page.indirectAction.page).toBe(emailOrPostPage)
    })

    it('should return an object with answer as an instance of CphKnownAnswer', () => {
      expect(page.indirectAction.answer).toBeInstanceOf(EmailOrPostAnswer)
    })

    it('should set the value to "email" in the answer', () => {
      expect(page.indirectAction.answer.toState()).toBe('email')
    })
  })

  describe('nextPage', () => {
    it('should return cphNumberPage for any value', () => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(EmailPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
