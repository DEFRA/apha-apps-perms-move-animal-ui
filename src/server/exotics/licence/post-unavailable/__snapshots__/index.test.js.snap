// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostUnavailablePage content should render expected response and content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">This service does not currently send licences by post</h1>
    </div>
  </div>
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">
            <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button" id="continue-button">
  Continue with email
</button>
        </div>
      </div>
    </form>
        "
`;
